lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -98,181 +128,461 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+        /// </summary>
+        /// <param name="failedEvaluator"></param>
+        public void OnNext(IFailedEvaluator failedEvaluator)
+        {
+            lock (_lock)
+            {
+                if (_taskManager.AreAllTasksCompleted())
+                {
+                    Logger.Log(Level.Verbose, "Evaluator with Id: {0} failed but IMRU task is completed. So ignoring.", failedEvaluator.Id);
+                    return;
+                }
+                Logger.Log(Level.Info, "Evaluator with Id: {0} failed with Exception: {1} in retry # {2}.", failedEvaluator.Id, failedEvaluator.EvaluatorException, _numberOfRetryForFaultTolerant);
+
+                var isMaster = _evaluatorManager.IsMasterEvaluatorId(failedEvaluator.Id);
+                _evaluatorManager.RecordFailedEvaluator(failedEvaluator.Id);
+                _contextManager.RemoveFailedContextInFailedEvaluator(failedEvaluator);
+
+                switch (_systemState.CurrentState)
+                {
+                    case SystemState.WaitingForEvaluator:
+                        if (IsRecoverable())
+                        {
+                            if (isMaster)
+                            {
+                                Logger.Log(Level.Info, "Requesting a master Evaluator.");
+                                _evaluatorManager.RequestUpdateEvaluator();
+                            }
+                            else
+                            {
+                                _serviceAndContextConfigurationProvider.RemoveEvaluatorIdFromPartitionIdProvider(failedEvaluator.Id);
+                                Logger.Log(Level.Info, "Requesting mapper Evaluators.");
+                                _evaluatorManager.RequestMapEvaluators(1);
+                            }
+                        }
+                        else
+                        {
+                            Logger.Log(Level.Warning, "The system is not recoverable, change the state to Fail.");
+                            _systemState.MoveNext(SystemStateEvent.NotRecoverable);
+                            FailAction();
+                        }
+                        break;
+
+                    case SystemState.SubmittingTasks:
+                    case SystemState.TasksRunning:
+                        // set system state to ShuttingDown
+                        _systemState.MoveNext(SystemStateEvent.FailedNode);
+                        _taskManager.RecordTaskFailWhenReceivingFailedEvaluator(failedEvaluator);
+                        _taskManager.CloseAllRunningTasks(TaskManager.CloseTaskByDriver);
+
+                        // Push evaluator id back to PartitionIdProvider if it is not master
+                        if (!isMaster)
+                        {
+                            _serviceAndContextConfigurationProvider.RemoveEvaluatorIdFromPartitionIdProvider(failedEvaluator.Id);
+                        }
+
+                        TryRecovery();
```

Comment:
This whole function is in lock and once you call StartAction() in TryRecovery() and request for more map evaluators, mapperCount there will always be one right? this is because no other IFaileEvaluator can be received in between because of lock.